index.js

```diff
@@ -0,0 +1,301 @@
+const BuiltInType = function(t){
+  const mapped = mapConstrToFn(t);
+
+  if (mapped !== t){
+    return mapped;
+  } else if (t.constructor === Function) {
+    return AutoPredicate(t);
+  } else {
+    return t;
+  }
+};
+
+const a = T.TypeVariable('a');
+
+const createIterator = function() {
+  return {
+    idx: 0,
+    val: this,
+    next() {
+
+      const keys = this.val._keys;
+
+      /* eslint-disable no-plusplus */
+      return this.idx === keys.length
+        ? {done: true}
+        : {value: this.val[keys[this.idx++]]};
+      /* eslint-enable no-plusplus */
+    },
+  };
+};
+
+const staticCase = function(options, b, ...args){
+  const f = options[b._name];
+  if (f){
+
+    const values = b._keys.map(k => b[k]);
+    return f(...[...values, ...args]);
+
+  } else if (options._) {
+    return options._(b);
+  } else {
+    // caseOn is untyped
+    // so this is possible
+    throw new TypeError(
+      'Non exhaustive case statement'
+    );
+  }
+};
+
+const CaseRecordType = function(keys, enums){
+  return T.RecordType(
+    keys.length
+    ? Object.assign(
+      ...keys.map(
+        k => ({
+          [k]: T.Function(values(enums[k]).concat(a)),
+        })
+      )
+    )
+    : {}
+  );
+};
+
+const ObjConstructorOf = prototype => (keys, name) => r =>
+  Object.assign(
+    Object.create(prototype),
+    r,
+    {
+      _keys: keys,
+      _name: name,
+      [Symbol.iterator]: createIterator,
+    }
+  );
+
+const RecursiveType = Type => v => typeof v === 'undefined' ? Type : v;
+
+const processRawCases =
+  (Type, rawCases) =>
+    map(
+      map(
+        B(
+          BuiltInType,
+          RecursiveType(Type)
+        )
+      ),
+      rawCases
+    );
+
+const CreateCaseConstructor = function(def, prototype, typeName, cases){
+
+  const objConstructorOf =
+    ObjConstructorOf(prototype);
+
+  return function createCaseConstructor(k){
+
+    const type = cases[k];
+
+    const isArray =
+      Array.isArray(type);
+
+    const keys = Object.keys(type);
+
+    const types =
+      isArray
+        ? type
+        : values(type);
+
+    const recordType =
+      isArray
+        ? T.RecordType(
+          zipObj(keys)(types)
+        )
+        : T.RecordType(type);
+
+    return {
+      [`${k}Of`]:
+        def(`${typeName}.${k}Of`,
+            {},
+            [recordType, recordType],
+            objConstructorOf(keys, k)),
+      [k]:
+        def(`${typeName}.${k}`,
+            {},
+            types.concat(recordType),
+            B(objConstructorOf(keys, k),
+              unapply(zipObj(keys)))),
+    };
+  };
+};
+
+
+const boundStaticCase = function(options){
+  return staticCase(options, this);
+};
+
+const Setup = function({check, ENV = T.env}){
+
+  const def =
+    T.create({
+      checkTypes: check,
+      env: ENV,
+    });
+
+  const CreateUnionType = function(typeName, rawCases, prototype = {}){
+
+    const Type = T.NullaryType(
+      typeName,
+      a => a && a['@@type'] === typeName
```

Comment:
Let's make the guard explicit by replacing `a &&` with `a != null &&`.

Let's use `x` as the parameter name to avoid shadowing our type variable named `a`.